infoschema/builder.go

```diff
@@ -309,6 +311,47 @@ func (b *Builder) applyReorganizePartition(m *meta.Meta, diff *model.SchemaDiff)
 	return tblIDs, nil
 }
 
+func (b *Builder) applyExchangeTablePartition(m *meta.Meta, diff *model.SchemaDiff) ([]int64, error) {
+	// The partitioned table is not affected until the last stage
+	if diff.OldTableID == diff.TableID && diff.OldSchemaID == diff.SchemaID {
+		return b.applyTableUpdate(m, diff)
+	}
+	ntSchemaID := diff.OldSchemaID
+	ntID := diff.OldTableID
+	ptSchemaID := diff.SchemaID
+	ptID := diff.TableID
+	if len(diff.AffectedOpts) > 0 {
```

Comment:
When we have `len(diff.AffectedOpts) > 0` for exchange partitions?

I found that `AffectedOpts` was introduced in https://github.com/pingcap/parser/pull/883 with https://github.com/pingcap/tidb/pull/17149. However, I cannot think of any case of `len(diff.AffectedOpts) > 0`.

/cc @crazycs520 @tiancaiamao

Reply:
This is in the case of upgrades, if there is an existing diff that is not yet processed, from an earlier version. Not sure the risk of that to happen though...
I manually tested it by reverting the `ddl/ddl_worker.go` file, which is where I changed/cleaned how to set the diff.